rfcs/0015-const-generics.md

```diff
@@ -0,0 +1,299 @@
+- RFC PR: [FuelLabs/sway-rfcs#42](https://github.com/FuelLabs/sway-rfcs/pull/42)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+Allows constant values as generic arguments.
+
+# Motivation
+
+[motivation]: #motivation
+
+Some types have constants, specifically unsigned integers, as their definition (e.g. arrays and string arrays). Without `const generics` it is impossible to have `impl` items for all instances of these types.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+`const generics` refer to the language syntax where generic parameters are defined as constant values, instead of types.
+
+A simple example would be:
+
+```rust
+fn id<const SIZE: usize>(array: [u64; SIZE]) -> [u64; SIZE] {
+    array
+}
+```
+
+This also allows `impl` items such as
+
+```rust
+impl<const N: usize> AbiEncode for str[N] {
+    ...
+}
+```
+
+This constant can be inferred or explicitly specified. When inferred, the syntax is no different than just using the item:
+
+```rust
+id([1u8])
+```
+
+In the example above, the Sway compiler will infer `SIZE` to be one, because `id` parameter will be infered to be `[u8; 1]`.
+
+For the cases where the compiler cannot infer this value, or this value comes from an expression, it is possible to do:
+
+```rust
+id::<1>([1u8]);
+id::<{1 + 1}>([1u8, 2u8]);
+```
+
+When the value is not a literal, but an expression, it is named "const generic expression" and it needs to be enclosed by curly braces. This will fail if the expression cannot be evaluated as `const`.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+This new syntax has three forms: declarations, instantiations and references.
+
+"const generics declarations" can appear anywhere all other generic arguments declarations are valid:
+
+1. Function/method declaration;
+1. Struct/Enum declaration;
+1. `impl` declarations.
+
+```rust
+// 1
+fn id<const N: usize>(...) { ... }
+
+// 2
+struct SpecialArray<const N: usize> {
+    inner: [u8; N],
+}
+
+//3
+impl<const N: usize> AbiEncode for str[N] {
+    ...
+}
+```
+
+"const generics instantiations" can appear anywhere all other generic argument instantiations are valid:
+
+1. Function/method reference;
+1. Struct/Enum reference;
+1. Fully qualified call paths.
+
+```rust
+// 1
+id::<1>([1u8]);
+special_array.do_something::<1>();
+
+// 2
+SpecialArray::<1>::new();
+
+// 3
+<SpecialArray::<1> as SpecialArrayTrait::<1>>::f();
+```
+
+"const generics references" can appear anywhere any other identifier appears. For semantic purposes, there is no difference between the reference of a "const generics" and a "normal" const.
+
+```rust
+fn f<const I: usize>() {
+    __log(I);
+}
+```
+
+Different from type generics, const generics cannot appear at:
+
+1. constraints;
+1. where types are expected.
+
+```rust
+// 1 - INVALID
+fn f<const I: usize>() where I > 10 {
```

Comment:
Supporting constraints can be added to future possibilities. One case I see as interesting, is ensuring that an array size is greater then zero. Or for small vec implementations to ensure the upper bound is not to high.